src/commonlib/include/commonlib/tcpa_log_serialized.h

```diff
@@ -6,12 +6,49 @@
 #include <stdint.h>
 
 #define MAX_TCPA_LOG_ENTRIES 50
-#define TCPA_DIGEST_MAX_LENGTH 64
 #define TCPA_PCR_HASH_NAME 50
 #define TCPA_PCR_HASH_LEN 10
 /* Assumption of 2K TCPA log size reserved for CAR/SRAM */
 #define MAX_PRERAM_TCPA_LOG_ENTRIES 15
 
+#if CONFIG(TPM_MEASURED_BOOT_SPEC_LOG)
+
+#define SHA1_DIGEST_LENGTH 20
+#define SHA256_DIGEST_LENGTH 32
+
+#define TCPA_DIGEST_MAX_LENGTH SHA256_DIGEST_LENGTH
+
+/* TCG_PCR_EVENT */
+struct tcpa_header {
+	uint32_t pcr;
+	uint32_t event_type;
+	uint8_t digest[SHA1_DIGEST_LENGTH];
+	uint32_t name_length;
+	char name[TCPA_PCR_HASH_NAME];
+} __packed;
+
+/* TCG_PCR_EVENT2 */
+struct tcpa_entry {
+	uint32_t pcr;
+	uint32_t event_type;
+	uint32_t digest_count;
+	uint16_t digest_type;
+	uint8_t digest[TCPA_DIGEST_MAX_LENGTH];
```

Comment:
This will break TCG format. `name_length` (or `EventSize`, or `eventDataSize`, depending on specification) should immediately follow last meaningful byte of digest, meaning this array will have different size based on digest algorithm. It is impossible to pack generalized form in C struct. We can either implement this with a walking pointer and append pairs of digest type and digest, or change the size of entry to SHA1 and put a warning that it is hardcoded and why. First option would be useful for TPM2.0.

Reply:
Thanks, I missed that. I found the right specification only at the very end, because TCG has the worst reference organization I've ever seen.